src/Particular.Analyzers/DiagnosticDescriptors.cs

```diff
@@ -4,204 +4,220 @@
             category: "Code",
             defaultSeverity: DiagnosticSeverity.Warning,
             isEnabledByDefault: true);
 
-        public static readonly DiagnosticDescriptor ImplicitCastFromDateTimeToDateTimeOffset = new DiagnosticDescriptor(
+        public static readonly DiagnosticDescriptor ImplicitCastFromDateTimeToDateTimeOffset = new(
             id: DiagnosticIds.ImplicitCastFromDateTimeToDateTimeOffset,
             title: "A DateTime should not be implicitly cast to a DateTimeOffset",
             messageFormat: "Do not implicitly cast a DateTime to a DateTimeOffset",
             category: "Code",
             defaultSeverity: DiagnosticSeverity.Warning,
             isEnabledByDefault: true);
 
-        public static readonly DiagnosticDescriptor NowUsedInsteadOfUtcNow = new DiagnosticDescriptor(
+        public static readonly DiagnosticDescriptor NowUsedInsteadOfUtcNow = new(
             id: DiagnosticIds.NowUsedInsteadOfUtcNow,
             title: "DateTime.UtcNow or DateTimeOffset.UtcNow should be used instead of DateTime.Now and DateTimeOffset.Now, unless the value is being used for displaying the current date-time in a user's local time zone",
             messageFormat: "Use {0}.UtcNow instead of {0}.Now",
             category: "Code",
             defaultSeverity: DiagnosticSeverity.Error,
             isEnabledByDefault: true);
 
-        public static readonly DiagnosticDescriptor NonInterfaceTypePrefixedWithI = new DiagnosticDescriptor(
+        public static readonly DiagnosticDescriptor NonInterfaceTypePrefixedWithI = new(
             id: DiagnosticIds.NonInterfaceTypePrefixedWithI,
             title: "A non-interface type should not be prefixed with I",
             messageFormat: "Remove the I prefix from {0}",
             category: "Code",
             defaultSeverity: DiagnosticSeverity.Warning,
             isEnabledByDefault: true);
 
-        public static readonly DiagnosticDescriptor StructuredLoggingWithRepeatedToken = new DiagnosticDescriptor(
+        public static readonly DiagnosticDescriptor DictionaryHasUnsupportedKeyType = new(
+            id: DiagnosticIds.DictionaryHasUnsupportedKeyType,
+            title: "Dictionary keys should implement IEquatable<T>",
+            messageFormat: "This {0} uses the type {1} as its key, which is a reference type that does not implement IEquatable<T>, which means that every object added to the collection will be evaluated using reference equality (object.Equals()) to be unique. This is often (but not always) a mistake, especially when used as a cache, because every lookup will be a cache miss. Don't blindly suppress this diagnostic or change this code. Be sure to consider whether the collection is really meant to evaluate equality using default reference equality, or if doing so will cause unintended side effects from objects that have the same data appearing to be duplicated.",
```

Comment:
```suggestion
            messageFormat: "This {0} uses the type {1} as its key, which is a reference type that does not implement IEquatable<T>, which means that every object added to the collection will be evaluated using reference equality (object.Equals()) to be unique. This is often (but not always) a mistake, especially when used as a cache, because every lookup will be a cache miss. Before suppressing this diagnostic or changing the code, be sure to consider whether the collection is really meant to evaluate equality using default reference equality, or if doing so will cause unintended side effects from objects that have the same data appearing to be duplicated.",
```